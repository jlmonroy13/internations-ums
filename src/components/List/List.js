import React from 'react';
import PropTypes from 'prop-types';
import { isLoaded } from 'react-redux-firebase';
import ListItem from '../ListItem/';
import EmptySection from '../EmptySection/';
import SelectForm from '../SelectForm/';
import EmptyItemsOnUms from '../EmptyItemsOnUms';

function List({
  type,
  title,
  entitiesObj,
  itemsObj,
  page,
  ownEntitiesObj,
  urlId,
  setMenuState,
  isMenuOpen,
}) {
  function onSetMenuState() {
    setMenuState(!isMenuOpen);
  }
  if ((!isLoaded(itemsObj) || !isLoaded(entitiesObj)) && itemsObj && entitiesObj) {
    return <div>Loading ...</div>;
  }
  if ((!itemsObj || !entitiesObj) && page !== 'details') return <EmptySection type={type} />;
  if (!ownEntitiesObj && !entitiesObj && page === 'details') return <EmptyItemsOnUms type={type} />;
  let items = itemsObj && entitiesObj ? Object.keys(entitiesObj).map(key => itemsObj[key]) : [];
  let itemsSelect = [];
  if (!ownEntitiesObj && entitiesObj && page === 'details') {
    itemsSelect = items;
    items = [];
  } else if (isLoaded(ownEntitiesObj) && page === 'details') {
    const ownEntitiesArr = Object.keys(ownEntitiesObj);
    itemsSelect = items.filter(item => ownEntitiesArr.indexOf(item.id) === -1);
    items = items.filter(item => ownEntitiesArr.indexOf(item.id) > -1);
  }
  return (
    <div>
      <h1 className="list__title">
        {title}
        {page !== 'details' ? (
          <span className="list__icon-plus icon icon--plus" onClick={onSetMenuState} />
        ) : null}
        {page === 'details' && itemsSelect.length ? (
          <SelectForm type={type} items={itemsSelect} urlId={urlId} />
        ) : null}
      </h1>
      <ul className="list">
        {items.reduce((acc, item) => {
          if (item && item.id) {
            acc.push(
              <ListItem
                key={item.id}
                data={item}
                type={type}
                page={page}
                urlId={urlId}
                itemsLength={items ? items.length : 0}
              />,
            );
          }
          return acc;
        }, [])}
      </ul>
    </div>
  );
}

List.propTypes = {
  type: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  page: PropTypes.string.isRequired,
  urlId: PropTypes.string.isRequired,
  itemsObj: PropTypes.object,
  entitiesObj: PropTypes.object,
  ownEntitiesObj: PropTypes.object,
  setMenuState: PropTypes.func,
  isMenuOpen: PropTypes.bool,
};

export default List;
