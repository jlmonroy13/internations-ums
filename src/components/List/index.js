import ListGroups from './ListGroupsContainer';
import ListUsers from './ListUsersContainer';
import ListGroupsOfUser from './ListGroupsOfUserContainer';
import ListUsersOfGroup from './ListUsersOfGroupContainer';

export {
  ListGroups,
  ListUsers,
  ListGroupsOfUser,
  ListUsersOfGroup,
};
