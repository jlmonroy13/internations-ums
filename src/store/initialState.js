export default {
  authorization: {
    uid: '',
    isLoading: true,
  },
  events: {
    editMode: false,
    isMenuOpen: false,
    isNavOpen: false,
  },
};
