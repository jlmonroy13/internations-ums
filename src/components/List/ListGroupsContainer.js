import { connect } from 'react-redux';
import { firebaseConnect, dataToJS, pathToJS } from 'react-redux-firebase';
import { setMenuState } from '../../actions/events';
import List from './List';

const fbWrappedComponent = firebaseConnect(['/groups', '/profile/groups'])(List);

function mapStateToProps(state) {
  const { firebase, events: { isMenuOpen } } = state;
  const entitiesObj = pathToJS(firebase, 'profile/groups');
  const itemsObj = dataToJS(firebase, 'groups');
  return {
    entitiesObj,
    itemsObj,
    isMenuOpen,
    urlId: '',
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setMenuState: state => dispatch(setMenuState(state)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(fbWrappedComponent);
