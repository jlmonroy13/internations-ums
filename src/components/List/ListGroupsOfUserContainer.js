import { connect } from 'react-redux';
import { firebaseConnect, dataToJS, pathToJS } from 'react-redux-firebase';
import List from './List';
import StrUtils from '../../utils/string';

const fbWrappedComponent = firebaseConnect(['/contacts', '/groups'])(List);

function mapStateToProps(state) {
  const { firebase, routing: { locationBeforeTransitions: { pathname: url } } } = state;
  const contactId = StrUtils.getStringBetween(url.slice(1), '/');
  const ownEntitiesObj = dataToJS(firebase, `contacts/${contactId}/group`);
  const entitiesObj = pathToJS(firebase, 'profile/groups');
  const itemsObj = dataToJS(firebase, 'groups');

  return {
    itemsObj,
    entitiesObj,
    ownEntitiesObj,
    urlId: contactId,
  };
}

export default connect(mapStateToProps)(fbWrappedComponent);
