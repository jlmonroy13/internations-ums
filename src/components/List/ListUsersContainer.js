import { connect } from 'react-redux';
import { firebaseConnect, dataToJS, pathToJS } from 'react-redux-firebase';
import { setMenuState } from '../../actions/events';
import List from './List';

const fbWrappedComponent = firebaseConnect(['/contacts'])(List);

function mapStateToProps(state) {
  const { firebase, events: { isMenuOpen } } = state;
  const entitiesObj = pathToJS(firebase, 'profile/contacts');
  const itemsObj = dataToJS(firebase, 'contacts');
  return {
    entitiesObj,
    itemsObj,
    isMenuOpen,
    urlId: '',
  };
}

function mapDispatchToProps(dispatch) {
  return {
    setMenuState: state => dispatch(setMenuState(state)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(fbWrappedComponent);
