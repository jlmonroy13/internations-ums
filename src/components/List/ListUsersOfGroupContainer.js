import { connect } from 'react-redux';
import { firebaseConnect, dataToJS, pathToJS } from 'react-redux-firebase';
import List from './List';
import StrUtils from '../../utils/string';

const fbWrappedComponent = firebaseConnect(['/contacts', '/groups'])(List);

function mapStateToProps(state) {
  const { firebase, routing: { locationBeforeTransitions: { pathname: url } } } = state;
  const groupId = StrUtils.getStringBetween(url.slice(1), '/');
  const ownEntitiesObj = dataToJS(firebase, `groups/${groupId}/user`);
  const entitiesObj = pathToJS(firebase, 'profile/contacts');
  const itemsObj = dataToJS(firebase, 'contacts');

  return {
    itemsObj,
    entitiesObj,
    ownEntitiesObj,
    urlId: groupId,
  };
}

export default connect(mapStateToProps)(fbWrappedComponent);
