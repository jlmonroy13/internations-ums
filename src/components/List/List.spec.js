import React from 'react';
import renderer from 'react-test-renderer';
import List from './List';
import { Provider } from 'react-redux';
import initialState from '../../store/initialState';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { getFirebase } from 'react-redux-firebase';

describe('<List />', () => {
  it('should render the list component', () => {
    const middlewares = [thunk.withExtraArgument(getFirebase)];
    const mockStore = configureStore(middlewares);
    const store = mockStore(initialState);
    const wrapper = renderer.create(
      <Provider store={store}>
        <List type="" title="" page="" urlId="" />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
