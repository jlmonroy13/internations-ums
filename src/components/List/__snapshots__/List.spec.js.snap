// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render the list component 1`] = `
<div
  className="details__warning"
>
  <p
    className="text--larger flush--bottom"
  >
    You don't have any  yet.
  </p>
  <p
    className="text--large"
  >
    Please add one.
  </p>
  <span
    className="details__warning-btn icon icon--plus"
    onClick={[Function]}
  />
</div>
`;
